File: articles/trust-cooperation-workplace-teams/python/trust_cooperation_model.py
"""
Trust and Cooperation in Workplace Teams
Advanced synthetic Python workflow.

Responsible-use scope:
This script is for synthetic-data research, methods demonstration,
institutional learning, trust-and-cooperation review, collaboration-system
education, reciprocity analysis, leadership and governance review, and
reproducible workflows. It is not an employee-screening, hiring, promotion,
compensation, discipline, termination, workplace surveillance, individual
performance-management, cultural-loyalty scoring, trustworthiness scoring,
cooperation scoring of workers, productivity-ranking, or psychological
assessment tool.
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, roc_auc_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler


ROOT = Path(__file__).resolve().parents[1]
RAW_DATA = ROOT / "data" / "raw" / "trust_cooperation_synthetic.csv"
TABLE_DIR = ROOT / "outputs" / "tables"
FIGURE_DIR = ROOT / "outputs" / "figures"

TABLE_DIR.mkdir(parents=True, exist_ok=True)
FIGURE_DIR.mkdir(parents=True, exist_ok=True)


FEATURES = [
    "competence_trust",
    "integrity_trust",
    "benevolence_trust",
    "reciprocity_norms",
    "communication_reliability",
    "opportunism_risk",
    "power_imbalance",
    "uncertainty_load",
    "leadership_credibility",
    "institutional_trust",
]


def load_data() -> pd.DataFrame:
    """Load synthetic trust-and-cooperation data."""
    return pd.read_csv(RAW_DATA)


def add_trust_scores(df: pd.DataFrame) -> pd.DataFrame:
    """Create cooperative capacity, trust risk, and review-priority scores."""
    scored = df.copy()

    scored["cooperative_capacity_score"] = (
        0.13 * scored["competence_trust"]
        + 0.14 * scored["integrity_trust"]
        + 0.12 * scored["benevolence_trust"]
        + 0.13 * scored["reciprocity_norms"]
        + 0.12 * scored["communication_reliability"]
        + 0.11 * scored["leadership_credibility"]
        + 0.11 * scored["institutional_trust"]
        - 0.10 * scored["opportunism_risk"]
        - 0.08 * scored["power_imbalance"]
        - 0.05 * scored["uncertainty_load"]
    ).clip(0, 100)

    scored["trust_risk_score"] = (
        0.10 * (100 - scored["competence_trust"])
        + 0.14 * (100 - scored["integrity_trust"])
        + 0.10 * (100 - scored["benevolence_trust"])
        + 0.12 * (100 - scored["reciprocity_norms"])
        + 0.10 * (100 - scored["communication_reliability"])
        + 0.10 * (100 - scored["leadership_credibility"])
        + 0.12 * (100 - scored["institutional_trust"])
        + 0.14 * scored["opportunism_risk"]
        + 0.10 * scored["power_imbalance"]
        + 0.08 * scored["uncertainty_load"]
    ).clip(0, 100)

    scored["reciprocity_gap_score"] = (
        0.35 * (100 - scored["reciprocity_norms"])
        + 0.25 * scored["power_imbalance"]
        + 0.20 * scored["opportunism_risk"]
        + 0.20 * (100 - scored["institutional_trust"])
    ).clip(0, 100)

    scored["trust_review_priority"] = np.select(
        [
            (scored["trust_risk_score"] >= 70)
            | (scored["cooperative_capacity_score"] < 45)
            | (scored["symbolic_cooperation_risk"] == 1),
            (scored["trust_risk_score"] >= 50)
            | (scored["cooperative_capacity_score"] < 60),
        ],
        ["Immediate Review", "Structured Review"],
        default="Routine Monitoring",
    )

    return scored


def summarize_teams(df: pd.DataFrame) -> pd.DataFrame:
    """Summarize trust-and-cooperation conditions by synthetic team."""
    return (
        df.groupby("team_id")
        .agg(
            observations=("period", "count"),
            avg_cooperative_capacity=("cooperative_capacity_score", "mean"),
            avg_trust_risk=("trust_risk_score", "mean"),
            avg_reciprocity_gap=("reciprocity_gap_score", "mean"),
            avg_competence_trust=("competence_trust", "mean"),
            avg_integrity_trust=("integrity_trust", "mean"),
            avg_benevolence_trust=("benevolence_trust", "mean"),
            avg_reciprocity_norms=("reciprocity_norms", "mean"),
            avg_communication_reliability=("communication_reliability", "mean"),
            avg_opportunism_risk=("opportunism_risk", "mean"),
            avg_power_imbalance=("power_imbalance", "mean"),
            avg_uncertainty_load=("uncertainty_load", "mean"),
            avg_leadership_credibility=("leadership_credibility", "mean"),
            avg_institutional_trust=("institutional_trust", "mean"),
            strong_collaboration_rate=("strong_collaboration", "mean"),
            symbolic_cooperation_risk_rate=("symbolic_cooperation_risk", "mean"),
        )
        .reset_index()
        .sort_values("avg_trust_risk", ascending=False)
    )


def fit_collaboration_model(df: pd.DataFrame) -> tuple[Pipeline, pd.DataFrame, str, float]:
    """Fit a synthetic strong-collaboration classifier."""
    x = df[FEATURES]
    y = df["strong_collaboration"]

    x_train, x_test, y_train, y_test = train_test_split(
        x, y, test_size=0.30, random_state=515, stratify=y
    )

    model = Pipeline(
        steps=[
            ("scale", StandardScaler()),
            ("logit", LogisticRegression(max_iter=3000)),
        ]
    )

    model.fit(x_train, y_train)

    probabilities = model.predict_proba(x_test)[:, 1]
    predictions = model.predict(x_test)

    auc = roc_auc_score(y_test, probabilities)
    report = classification_report(y_test, predictions)

    coefficients = pd.DataFrame(
        {
            "feature": FEATURES,
            "coefficient": model.named_steps["logit"].coef_[0],
        }
    ).sort_values("coefficient", ascending=False)

    return model, coefficients, report, auc


def build_scenario_table(model: Pipeline) -> pd.DataFrame:
    """Compare high-trust and low-trust synthetic collaboration scenarios."""
    scenarios = pd.DataFrame(
        [
            {
                "scenario": "Trust-supported cooperative team",
                "competence_trust": 84,
                "integrity_trust": 82,
                "benevolence_trust": 78,
                "reciprocity_norms": 81,
                "communication_reliability": 83,
                "opportunism_risk": 16,
                "power_imbalance": 20,
                "uncertainty_load": 52,
                "leadership_credibility": 84,
                "institutional_trust": 82,
            },
            {
                "scenario": "Low-trust symbolic cooperation risk",
                "competence_trust": 43,
                "integrity_trust": 36,
                "benevolence_trust": 34,
                "reciprocity_norms": 39,
                "communication_reliability": 41,
                "opportunism_risk": 72,
                "power_imbalance": 69,
                "uncertainty_load": 52,
                "leadership_credibility": 34,
                "institutional_trust": 32,
            },
        ]
    )

    scenarios["predicted_strong_collaboration_probability"] = model.predict_proba(
        scenarios[FEATURES]
    )[:, 1]

    scenarios["trust_risk_score"] = (
        0.10 * (100 - scenarios["competence_trust"])
        + 0.14 * (100 - scenarios["integrity_trust"])
        + 0.10 * (100 - scenarios["benevolence_trust"])
        + 0.12 * (100 - scenarios["reciprocity_norms"])
        + 0.10 * (100 - scenarios["communication_reliability"])
        + 0.10 * (100 - scenarios["leadership_credibility"])
        + 0.12 * (100 - scenarios["institutional_trust"])
        + 0.14 * scenarios["opportunism_risk"]
        + 0.10 * scenarios["power_imbalance"]
        + 0.08 * scenarios["uncertainty_load"]
    ).clip(0, 100)

    return scenarios


def main() -> None:
    df = load_data()
    scored = add_trust_scores(df)

    scored.to_csv(TABLE_DIR / "scored_trust_cooperation_observations.csv", index=False)

    team_summary = summarize_teams(scored)
    team_summary.to_csv(TABLE_DIR / "team_trust_cooperation_summary.csv", index=False)

    model, coefficients, report, auc = fit_collaboration_model(scored)
    coefficients.to_csv(TABLE_DIR / "strong_collaboration_coefficients.csv", index=False)

    scenario_table = build_scenario_table(model)
    scenario_table.to_csv(TABLE_DIR / "trust_cooperation_scenario_comparison.csv", index=False)

    with open(TABLE_DIR / "strong_collaboration_model_report.txt", "w", encoding="utf-8") as handle:
        handle.write("Synthetic trust and cooperation strong-collaboration model\n")
        handle.write("=========================================================\n\n")
        handle.write(f"AUC: {auc:.3f}\n\n")
        handle.write(report)
        handle.write(
            "\n\nResponsible-use reminder: This model is a synthetic methods "
            "demonstration for institutional learning and trust-and-cooperation "
            "review. It must not be used for employee screening, employment "
            "selection, workplace surveillance, individual performance management, "
            "cultural loyalty scoring, trustworthiness scoring, cooperation scoring "
            "of workers, productivity ranking, or psychological assessment.\n"
        )

    print("Wrote outputs to:", TABLE_DIR)
    print(team_summary)
    print(scenario_table)


if __name__ == "__main__":
    main()

File: articles/trust-cooperation-workplace-teams/r/trust_cooperation_model.R
# Trust and Cooperation in Workplace Teams
# Advanced synthetic R workflow
#
# Responsible-use scope:
# This script is for synthetic-data research, methods demonstration,
# institutional learning, trust-and-cooperation review, collaboration-system
# education, reciprocity analysis, leadership and governance review, and
# reproducible workflows. It is not an employee-screening, hiring, promotion,
# compensation, discipline, termination, workplace surveillance, individual
# performance-management, cultural-loyalty scoring, trustworthiness scoring,
# cooperation scoring of workers, productivity-ranking, or psychological
# assessment tool.

suppressPackageStartupMessages({
  library(dplyr)
  library(readr)
  library(ggplot2)
})

root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."), mustWork = FALSE)
raw_path <- file.path(root, "data", "raw", "trust_cooperation_synthetic.csv")
table_dir <- file.path(root, "outputs", "tables")
figure_dir <- file.path(root, "outputs", "figures")

dir.create(table_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(figure_dir, recursive = TRUE, showWarnings = FALSE)

trust_data <- read_csv(raw_path, show_col_types = FALSE)

scored <- trust_data %>%
  mutate(
    cooperative_capacity_score = pmin(
      pmax(
        0.13 * competence_trust +
          0.14 * integrity_trust +
          0.12 * benevolence_trust +
          0.13 * reciprocity_norms +
          0.12 * communication_reliability +
          0.11 * leadership_credibility +
          0.11 * institutional_trust -
          0.10 * opportunism_risk -
          0.08 * power_imbalance -
          0.05 * uncertainty_load,
        0
      ),
      100
    ),
    trust_risk_score = pmin(
      pmax(
        0.10 * (100 - competence_trust) +
          0.14 * (100 - integrity_trust) +
          0.10 * (100 - benevolence_trust) +
          0.12 * (100 - reciprocity_norms) +
          0.10 * (100 - communication_reliability) +
          0.10 * (100 - leadership_credibility) +
          0.12 * (100 - institutional_trust) +
          0.14 * opportunism_risk +
          0.10 * power_imbalance +
          0.08 * uncertainty_load,
        0
      ),
      100
    ),
    reciprocity_gap_score = pmin(
      pmax(
        0.35 * (100 - reciprocity_norms) +
          0.25 * power_imbalance +
          0.20 * opportunism_risk +
          0.20 * (100 - institutional_trust),
        0
      ),
      100
    ),
    trust_review_priority = case_when(
      trust_risk_score >= 70 |
        cooperative_capacity_score < 45 |
        symbolic_cooperation_risk == 1 ~ "Immediate Review",
      trust_risk_score >= 50 |
        cooperative_capacity_score < 60 ~ "Structured Review",
      TRUE ~ "Routine Monitoring"
    )
  )

write_csv(scored, file.path(table_dir, "r_scored_trust_cooperation_observations.csv"))

team_summary <- scored %>%
  group_by(team_id) %>%
  summarise(
    observations = n(),
    avg_cooperative_capacity = mean(cooperative_capacity_score),
    avg_trust_risk = mean(trust_risk_score),
    avg_reciprocity_gap = mean(reciprocity_gap_score),
    avg_competence_trust = mean(competence_trust),
    avg_integrity_trust = mean(integrity_trust),
    avg_benevolence_trust = mean(benevolence_trust),
    avg_reciprocity_norms = mean(reciprocity_norms),
    avg_communication_reliability = mean(communication_reliability),
    avg_opportunism_risk = mean(opportunism_risk),
    avg_power_imbalance = mean(power_imbalance),
    avg_uncertainty_load = mean(uncertainty_load),
    avg_leadership_credibility = mean(leadership_credibility),
    avg_institutional_trust = mean(institutional_trust),
    strong_collaboration_rate = mean(strong_collaboration),
    symbolic_cooperation_risk_rate = mean(symbolic_cooperation_risk),
    .groups = "drop"
  ) %>%
  arrange(desc(avg_trust_risk))

write_csv(team_summary, file.path(table_dir, "r_team_trust_cooperation_summary.csv"))

collaboration_model <- glm(
  strong_collaboration ~ competence_trust + integrity_trust +
    benevolence_trust + reciprocity_norms + communication_reliability +
    opportunism_risk + power_imbalance + uncertainty_load +
    leadership_credibility + institutional_trust,
  family = binomial(),
  data = scored
)

capture.output(
  summary(collaboration_model),
  file = file.path(table_dir, "r_strong_collaboration_model_summary.txt")
)

risk_plot <- ggplot(
  team_summary,
  aes(x = reorder(team_id, avg_trust_risk), y = avg_trust_risk)
) +
  geom_col() +
  coord_flip() +
  labs(
    title = "Synthetic Trust and Cooperation Risk by Team",
    x = "Synthetic team",
    y = "Average trust risk score"
  ) +
  theme_minimal()

ggsave(
  filename = file.path(figure_dir, "r_trust_cooperation_risk_by_team.png"),
  plot = risk_plot,
  width = 8,
  height = 5,
  dpi = 160
)

integrity_plot <- ggplot(
  scored,
  aes(x = integrity_trust, y = cooperative_capacity_score)
) +
  geom_point(alpha = 0.55) +
  geom_smooth(method = "lm", se = TRUE) +
  labs(
    title = "Integrity Trust and Cooperative Capacity",
    x = "Integrity trust",
    y = "Cooperative capacity score"
  ) +
  theme_minimal()

ggsave(
  filename = file.path(figure_dir, "r_integrity_trust_cooperative_capacity.png"),
  plot = integrity_plot,
  width = 8,
  height = 5,
  dpi = 160
)

print(team_summary)
message("Responsible-use reminder: synthetic trust and cooperation workflow only.")
